lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs

```diff
@@ -452,49 +549,76 @@ withBFTNode
     -> NodeParams
     -- ^ Parameters used to generate config files.
     -> (FilePath -> Block -> (NetworkParameters, NodeVersionData) -> IO a)
     -- ^ Callback function with genesis parameters
     -> IO a
-withBFTNode tr baseDir (NodeParams severity systemStart (port, peers)) action =
+withBFTNode
+    tr
+    baseDir
+    i
+    (NodeParams severity systemStart mode (port, peers))
+    action =
     bracketTracer' tr "withBFTNode" $ do
-        createDirectory dir
+        createDirectoryIfMissing False dir
+
+        [bftKey, bftDelegCert] <- forM
+            ["delegate-keys.00" <> show i <> ".key", "delegation-cert.00" <> show i <> ".json"]
+            (\f -> copyFile (source </> f) (dir </> f) $> (dir </> f))
 
+        -- TODO: Doesn't support indexing by i
         [vrfPrv, kesPrv, opCert] <- forM
-            ["bft-leader.vrf.skey", "bft-leader.kes.skey", "bft-leader.opcert"]
+            [ "bft-leader" <> ".vrf.skey"
+            , "bft-leader" <> ".kes.skey"
+            , "bft-leader" <> ".opcert"
+            ]
             (\f -> copyFile (source </> f) (dir </> f) $> (dir </> f))
 
         (config, block0, networkParams, versionData)
-            <- genConfig dir severity systemStart
+            <- genConfig dir severity systemStart mode
         topology <- genTopology dir peers
 
-        let cfg = CardanoNodeConfig
+
+        let byronKeys x =
+                    x { nodeDlgCertFile = Just bftDelegCert
+                      , nodeSignKeyFile = Just bftKey
+                      }
+        let shelleyKeys x =
+                    x { nodeOpCertFile = Just opCert
+                      , nodeKesKeyFile = Just kesPrv
+                      , nodeVrfKeyFile = Just vrfPrv
+                      }
+
+        let makeModeSpecific x = case mode of
+                 RunByronShelley _ -> byronKeys . shelleyKeys $ x
+                 RunByron -> byronKeys x
+                 RunShelley -> shelleyKeys x
+
+        let cfg = makeModeSpecific $ CardanoNodeConfig
                 { nodeDir = dir
                 , nodeConfigFile = config
                 , nodeTopologyFile = topology
                 , nodeDatabaseDir = "db"
                 , nodeDlgCertFile = Nothing
                 , nodeSignKeyFile = Nothing
-                , nodeOpCertFile = Just opCert
-                , nodeKesKeyFile = Just kesPrv
-                , nodeVrfKeyFile = Just vrfPrv
+                , nodeOpCertFile = Nothing
+                , nodeKesKeyFile = Nothing
+                , nodeVrfKeyFile = Nothing
                 , nodePort = Just (NodePort port)
                 , nodeLoggingHostname = Just name
                 }
 
         withCardanoNodeProcess tr name cfg $ \(CardanoNodeConn socket) -> do
-            setEnv "CARDANO_NODE_SOCKET_PATH" socket
-            (rawTx, faucetPrv) <- prepareKeyRegistration tr dir
-            tx <- signTx tr dir rawTx [faucetPrv]
-            submitTx tr "pre-registered stake key" tx
```

Comment:
This is important and shouldn't be removed. I get that it can only happen after the hard-fork, but I don't see it moved at the relevant place.

Reply:
Yes, you're right.

I mentioned it in the PR description as a remaining todo:
> Todo clean up faucet config/code, and make sure special wallets are included